tfx/extensions/google_cloud_kubernetes/runner.py

```diff
@@ -0,0 +1,245 @@
+  for i in range(num_workers):
+    tf_config = json.dumps({
+        'cluster': {
+            'worker': worker_hosts
+        },
+        'task': {'type': 'worker', 'index': i}
+    })
+    pod = client.V1Pod(
+        metadata=client.V1ObjectMeta(
+            name=pod_names[i],
+            labels={
+                'name': 'training',
+                'id': unique_id,
+                'task': str(i),
+            },
+        ),
+        spec=client.V1PodSpec(
+            containers=[
+                client.V1Container(
+                    name='worker-pod',
+                    image=tfx_image,
+                    command=_COMMAND,
+                    args=job_args,
+                    security_context=client.V1SecurityContext(
+                        privileged=True,
+                    ),
+                    env=[
+                        client.V1EnvVar(
+                            name='TF_CONFIG',
+                            value=tf_config,
+                        ),
+                    ],
+                    ports=[
+                        client.V1ContainerPort(
+                            container_port=5000,
+                        ),
+                    ],
+                    resources=client.V1ResourceRequirements(
+                        limits={
+                            'nvidia.com/gpu': num_gpus_per_worker,
+                        },
+                    ) if num_gpus_per_worker > 0 else None,
+                ),
+            ],
+            restart_policy=kube_utils.RestartPolicy.NEVER.value,
+        ),
+    )
+    try:
+      api_instance.create_namespaced_pod(namespace='default', body=pod)
+    except ApiException as e:
+      logging.error(
+          'Exception when calling CoreV1Api->create_namespaced_pod: %s' % e)
+
+  logging.info('created {} worker pods'.format(num_workers))
+
+
+def create_worker_services(training_inputs: Dict[Text, Any],
+                           unique_id: Text):
+  """Create worker services for multi-worker training."""
+  num_workers = training_inputs.get('num_workers', 1)
+  service_names = _build_service_names(num_workers=num_workers,
+                                       unique_id=unique_id)
+  api_instance = kube_utils.make_core_v1_api()
+
+  # TODO(ericlege): consider using a jinja2 template instead
+  for i in range(num_workers):
+    service = client.V1Service(
+        metadata=client.V1ObjectMeta(
+            name=service_names[i],
+        ),
+        spec=client.V1ServiceSpec(
+            selector={
+                'name': 'training',
+                'id': unique_id,
+                'task': str(i),
+            },
+            ports=[
+                client.V1ServicePort(
+                    port=5000,
+                ),
+            ],
+        ),
+    )
+    try:
+      api_instance.create_namespaced_service(namespace='default', body=service)
+    except ApiException as e:
+      logging.error(
```

Comment:
Ditto: return after logging? Should this be a fatal instead / should we reraise the exception?